kconfig: doc: Split up symbol reference by path

ulfalizer opened the pull request:
Main commit:

```
kconfig: doc: Split up symbol reference by path

Use --modules to split the Kconfig reference up by where symbols are
defined. See the argparse docstring for --modules in
doc/scripts/genrest.py.

Not sure what the best way to split things up is, so feedback would be
appreciated. I just pulled out some top-level directories. It could
always be tweaked later.

If a symbol is defined in more than one module (by being defined in
multiple locations), it appears on multiple index pages.

To build the documentation, do

    $ mkdir doc/b && cd doc/b
    $ cmake -GNinja ..
    $ ninja htmldocs
    (output in html/reference/kconfig/index.html)
```

Depends on this script change, which is also included in this PR:

```
genrest: List all symbols on main index page and refactor

Instead of having index.rst just link to other index pages in --modules
mode, list all symbols on it, and have links to the module index pages
at the top.

Get rid of index-all.rst and index-main.rst (and all related options).
index-main.rst is no longer needed, and index-other.rst (symbols outside
--modules) might not be that useful to people reading the documentation
(and could be added back if needed).

Also refactor and streamline the code a bunch. For the main index page,
the only difference between modules and non-modules mode is now whether
there's links to other index pages at the top.

Suggested by David B. Kinder.
```

Also includes this tweak to show help texts on the index page:

```
genrest: Show symbol help texts on index page

Instead of showing the prompt for symbols on the index page, show their
help texts. This makes searching easier.

Fall back on the prompt if no help text is available.

Also change the code to only show one of the prompts if several are
available (happens if a symbol is defined in multiple locations and adds
a prompt in more than one of them). It's probably overkill to show them
all, and it doesn't come up that often.

Suggested by David B. Kinder.
```

ulfalizer commented:
Screenshot for the configuration in this PR:

![Skärmbild från 2019-11-04 20-22-59](https://user-images.githubusercontent.com/154768/68150927-039ba580-ff41-11e9-9e98-d06fba9e2a3b.png)

 Each link leads to an index of symbols similar to the old index that listed everything.

ulfalizer commented:
Not sure whether the path thing is flexible enough either. This depends on symbols fitting neatly into paths depending on category too.

SebastianBoe commented:
Could you explain why we want this change?

I don't use this web page, so I don't know if this change is aligned with what the users want or not. But I could imagine that some users use it as a searchable index, in which case having all the options on the same page is necessary.

ulfalizer commented:
> Could you explain why we want this change?
> 
> I don't use this web page, so I don't know if this change is aligned with what the users want or not. But I could imagine that some users use it as a searchable index, in which case having all the options on the same page is necessary.

To get a better overview. We have thousands of symbols, and if someone just wants to check the core kernel configuration options, there's no easy way to do it at the moment.

Note that there's also a page that lists all symbols (the `All Configuration Options` link).

SebastianBoe commented:
> Note that there's also a page that lists all symbols (the All Configuration Options link).

I was not aware, then this LGTM.

SebastianBoe commented on doc/CMakeLists.txt:
```diff
@@ -204,7 +204,16 @@ add_custom_target(
   KCONFIG_TURBO_MODE=${KCONFIG_TURBO_MODE}
   KCONFIG_DOC_MODE=1
-  ${PYTHON_EXECUTABLE} scripts/genrest.py ${RST_OUT}/doc/reference/kconfig/
+    ${PYTHON_EXECUTABLE} scripts/genrest.py ${RST_OUT}/doc/reference/kconfig/
```
something is off with indentation

ulfalizer replied:
Was deliberate, to make the command part stand out better.

carlescufi commented:
@dbkinder @nashif thoughts on this? I think it's a definite improvement in terms of user experience, but I would place the "All configuration options" entry at the top and clearly separated.

dbkinder commented:
You can view the build artifacts for this documentation change to see how it actualy looks, here: https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/20322/docs/reference/kconfig/index.html

dbkinder commented:
I agree that the "All Configuration Options" needs to be more visible.  The current version of this configuration page (https://docs.zephyrproject.org/latest/reference/kconfig/index.html) has the long list of symbols on the page itself.  Perhaps your change can keep that, but add the subsection links as a new section above them, and change the section heading to "All Configuration Symbols", something like:
![image](https://user-images.githubusercontent.com/11063618/68514832-b6b31a00-0233-11ea-81a5-76f8b48f4358.png)

dbkinder commented:
Separately, we need to work on that introductory text on the page :)

nashif commented:
why do we need the "All" option? Dont we already have coverage in the other modules?

ulfalizer commented:
@dbkinder 
Yeah, something like that might be nice. Sorry, didn't see the comment until now.

Marking DNM for now as I'm busy with some other stuff, but will return to this later.

dbkinder commented:
> why do we need the "All" option? Dont we already have coverage in the other modules?

It's convenient to have a page with all symbols for CTRL-F searching, as @SebastianBoe points out.

dbkinder commented:
@ulfalizer Not sure I can comment on the specific module groupings, but there does need to be an easier way for a developer to know what configuration options they can and should tweak.  Need some input from the developers to know more, but your grouping makes sense to me.

On a separate note and maybe better as a separate issue/PR, I find the descriptions of the symbols sorta silly and are often just a paraphrase of the symbol name: 

    CONFIG_2ND_LEVEL_INTERRUPTS              Second-level interrupt support

I tried a tweak to the genrest.py to instead put the ``.help`` content as the description (instead of the ``.prompt``) and it seems a more useful symbol index, don't you think:

![Screenshot from 2019-11-20 16-33-56](https://user-images.githubusercontent.com/11063618/69290194-fecb2880-0bb3-11ea-9f86-f2e7c6d7abae.png)

Something like this (needed to add the indent for each line of the help text to keep Sphinx happy):
```
332         rst += "   * - :option:`CONFIG_{}`\n     - {}\n".format(
333             sym.name, " / ".join(node.help.replace("\n","\n       ") for node in sym.nodes if node.help))
```
and also tweaked the column width percentage for this table with:
```
322    :widths: 30 70
```

ulfalizer commented:
> @ulfalizer Not sure I can comment on the specific module groupings, but there does need to be an easier way for a developer to know what configuration options they can and should tweak. Need some input from the developers to know more, but your grouping makes sense to me.

One thing I've been thinking of is having a checkbox like *Show invisible symbols* that'd be off by default. When off, only symbols with prompts would be shown. Those are the symbols that show up in `menuconfig` and can be changed by the user.

It'd probably require JavaScript though, and some way to mark up rows in the `.rst` output so that the JavaScript code can know which rows have invisible symbols on them.

> I tried a tweak to the genrest.py to instead put the `.help` content as the description (instead of the `.prompt`) and it seems a more useful symbol index, don't you think:
> 
> ![Screenshot from 2019-11-20 16-33-56](https://user-images.githubusercontent.com/11063618/69290194-fecb2880-0bb3-11ea-9f86-f2e7c6d7abae.png)
> 
> Something like this (needed to add the indent for each line of the help text to keep Sphinx happy). Might be better to only show the first .help if there's more than one, but you'll know better:
> 
> ```
> 332         rst += "   * - :option:`CONFIG_{}`\n     - {}\n".format(
> 333             sym.name, " / ".join(node.help.replace("\n","\n       ") for node in sym.nodes if node.help))
> ```
> 
> and also tweaked the column width percentage for this table with:
> 
> ```
> 322    :widths: 30 70
> ```

Ohh... that seems pretty nice. Maybe the prompt could be used as a fallback too when there's no help text.

Only downside I can think of is that it might make the index page even heavier, though I haven't treid it out.

I'll start with changing the index page to match https://github.com/zephyrproject-rtos/zephyr/pull/20322#issuecomment-552015239.

nashif commented:
> It's convenient to have a page with all symbols for CTRL-F searching, as @SebastianBoe points out.

we have a search functionality for this already.

ulfalizer commented:
> > It's convenient to have a page with all symbols for CTRL-F searching, as @SebastianBoe points out.
> 
> we have a search functionality for this already.

`Ctrl-F` is nicer though, IMO.

Wonder if we could add one of those search-with-Google thingies.

ulfalizer committed:
genrest: List all symbols on main index page and refactor

Instead of having index.rst just link to other index pages in --modules
mode, list all symbols on it, and have links to the module index pages
at the top.

Get rid of index-all.rst and index-main.rst (and all related options).
index-all.rst is no longer needed, and index-main.rst (symbols outside
--modules) might not be that useful to people reading the documentation
(and could be added back if needed).

Also refactor and streamline the code a bunch. For the main index page,
the only difference between modules and non-modules mode is now whether
there's links to other index pages at the top.

Suggested by David B. Kinder.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
kconfig: doc: Split up symbol reference by path

Use --modules to split the Kconfig reference up by where symbols are
defined. See the argparse docstring for --modules in
doc/scripts/genrest.py.

Not sure what the best way to split things up is, so feedback would be
appreciated. I just pulled out some top-level directories. It could
always be tweaked later.

If a symbol is defined in more than one module (by being defined in
multiple locations), it appears on multiple index pages.

To build the documentation, do

    $ mkdir doc/b && cd doc/b
    $ cmake -GNinja ..
    $ ninja htmldocs
    (output in html/reference/kconfig/index.html)

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
@dbkinder @SebastianBoe 
Now updated to put all symbols on the main index page. See the updated PR description.

Shouldn't be a regression to merge this now at least, since the main index page is easily searchable like before, so I removed the DNM tag. I think the --modules list could be tweaked as we go.

Will play around a bit with putting the help texts onto the index page too, like @dbkinder suggested.

ulfalizer committed:
genrest: Show symbol help texts on index page

Instead of showing the prompt for symbols on the index page, show their
help texts. This makes searching easier.

Fall back on the prompt if no help text is available.

Also change the code to only show one of the prompts if several are
available (happens if a symbol is defined in multiple locations and adds
a prompt in more than one of them). It's probably overkill to show them
all, and it doesn't come up that often.

Suggested by David B. Kinder.

Co-authored-by: David B. Kinder <david.b.kinder@intel.com>
Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
@dbkinder 
Added the index page help text tweak now too, in a separate commit.

dbkinder reviewed (changes requested):
Looks great, with one tweak to the column widths on the table...

dbkinder commented:
I also verified that ``make htmldocs-fast`` still works :)

ulfalizer commented on doc/scripts/genrest.py:
```diff
@@ -248,128 +205,133 @@ def parse_args():
+
+    rst = "\n\n" + title + "\n" + len(title)*"*" + """
+
+.. list-table::
```
Fixed... knew there was something I had forgotten.

dbkinder reviewed (approved):
Nice, thanks!
